pkg/scale/decode.go

```diff
@@ -0,0 +1,539 @@
+func (ds *decodeState) decodeSmallInt(firstByte, mode byte) (out int64, err error) {
+	switch mode {
+	case 0:
+		out = int64(firstByte >> 2)
+	case 1:
+		var buf byte
+		buf, err = ds.ReadByte()
+		if err != nil {
+			break
+		}
+		out = int64(binary.LittleEndian.Uint16([]byte{firstByte, buf}) >> 2)
+	case 2:
+		buf := make([]byte, 3)
+		_, err = ds.Read(buf)
+		if err != nil {
+			break
+		}
+		out = int64(binary.LittleEndian.Uint32(append([]byte{firstByte}, buf...)) >> 2)
+	}
+	return
+}
+
+// decodeBigInt decodes a SCALE encoded byte array into a *big.Int
+// Works for all integers, including ints > 2**64
+func (ds *decodeState) decodeBigInt(dstv reflect.Value) (err error) {
+	b, err := ds.ReadByte()
+	if err != nil {
+		return
+	}
+
+	var output *big.Int
+	// check mode of encoding, stored at 2 least significant bits
+	mode := b & 0x03
+	switch {
+	case mode <= 2:
+		var tmp int64
+		tmp, err = ds.decodeSmallInt(b, mode)
+		if err != nil {
+			break
+		}
+		output = big.NewInt(tmp)
+
+	default:
+		// >4 byte mode
+		topSixBits := b >> 2
+		byteLen := uint(topSixBits) + 4
+
+		buf := make([]byte, byteLen)
+		_, err = ds.Read(buf)
+		if err != nil {
+			err = fmt.Errorf("could not decode invalid big.Int: %v", err)
+			break
+		}
+		o := reverseBytes(buf)
+		output = big.NewInt(0).SetBytes(o)
+	}
+	dstv.Set(reflect.ValueOf(output))
+	return
+}
+
+// decodeFixedWidthInt decodes integers < 2**32 by reading the bytes in little endian
+func (ds *decodeState) decodeFixedWidthInt(dstv reflect.Value) (err error) {
+	in := dstv.Interface()
+	var out interface{}
+	switch in.(type) {
+	case int8:
+		var b byte
+		b, err = ds.ReadByte()
+		if err != nil {
+			break
+		}
+		out = int8(b)
+	case uint8:
+		var b byte
+		b, err = ds.ReadByte()
+		if err != nil {
+			break
+		}
+		out = uint8(b)
+	case int16:
+		buf := make([]byte, 2)
+		_, err = ds.Read(buf)
+		if err != nil {
+			break
+		}
+		out = int16(binary.LittleEndian.Uint16(buf))
+	case uint16:
+		buf := make([]byte, 2)
+		_, err = ds.Read(buf)
+		if err != nil {
+			break
+		}
+		out = binary.LittleEndian.Uint16(buf)
+	case int32:
+		buf := make([]byte, 4)
+		_, err = ds.Read(buf)
+		if err != nil {
+			break
+		}
+		out = int32(binary.LittleEndian.Uint32(buf))
+	case uint32:
+		buf := make([]byte, 4)
+		_, err = ds.Read(buf)
+		if err != nil {
+			break
+		}
+		out = binary.LittleEndian.Uint32(buf)
+	case int64:
+		buf := make([]byte, 8)
+		_, err = ds.Read(buf)
+		if err != nil {
+			break
+		}
+		out = int64(binary.LittleEndian.Uint64(buf))
+	case uint64:
+		buf := make([]byte, 8)
+		_, err = ds.Read(buf)
+		if err != nil {
+			break
+		}
+		out = binary.LittleEndian.Uint64(buf)
+	case int:
+		buf := make([]byte, 8)
+		_, err = ds.Read(buf)
+		if err != nil {
+			break
+		}
+		out = int(binary.LittleEndian.Uint64(buf))
+	case uint:
+		buf := make([]byte, 8)
+		_, err = ds.Read(buf)
+		if err != nil {
+			break
+		}
+		out = uint(binary.LittleEndian.Uint64(buf))
```

Comment:
are these cases needed still?

Reply:
there's a performance benefit if it falls into these cases, but technically no.  The `default` case that uses reflection will work in all these cases.

Reply:
nope